### PR TITLE
OCPQE-17804: add oc-mirror payload image step

### PR DESCRIPTION
**Pull request by jianlinliu**
Follow up https://github.com/openshift/release/pull/46249

### CODE REVIEW

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-arm64-stable-aws-ipi-disc-priv-sno-by-oc-mirror-f14

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-arm64-stable-aws-ipi-disc-priv-sno-by-oc-mirror-f14

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-arm64-stable-aws-ipi-disc-priv-sno-by-oc-mirror-f14

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-arm64-stable-aws-ipi-disc-priv-sno-by-oc-mirror-f14

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-arm64-stable-aws-ipi-disc-priv-sno-by-oc-mirror-f14

**Comment by jianlinliu** [comment]
pj-rehearse job failed, seen from installer gather log:
```
Jul 05 06:13:44 ip-10-0-49-25 release-image-download.sh[2156]: time="2024-07-05T06:13:44Z" level=warning msg="Failed, retrying in 1s ... (1/3). Error: initializing source docker://registry.build05.ci.openshift.org/ci-op-vljni94d/release@sha256:250b9972d03463ba38a85b657066aa0706bf348608797aa50ddda769d4afdf27: pinging container registry registry.build05.ci.openshift.org: Get \"https://registry.build05.ci.openshift.org/v2/\": dial tcp 54.146.60.183:443: i/o timeout"
Jul 05 06:14:45 ip-10-0-49-25 release-image-download.sh[2156]: time="2024-07-05T06:14:45Z" level=warning msg="Failed, retrying in 1s ... (2/3). Error: initializing source docker://registry.build05.ci.openshift.org/ci-op-vljni94d/release@sha256:250b9972d03463ba38a85b657066aa0706bf348608797aa50ddda769d4afdf27: pinging container registry registry.build05.ci.openshift.org: Get \"https://registry.build05.ci.openshift.org/v2/\": dial tcp 54.146.60.183:443: i/o timeout"
Jul 05 06:15:46 ip-10-0-49-25 release-image-download.sh[2156]: time="2024-07-05T06:15:46Z" level=warning msg="Failed, retrying in 1s ... (3/3). Error: initializing source docker://registry.build05.ci.openshift.org/ci-op-vljni94d/release@sha256:250b9972d03463ba38a85b657066aa0706bf348608797aa50ddda769d4afdf27: pinging container registry registry.build05.ci.openshift.org: Get \"https://registry.build05.ci.openshift.org/v2/\": dial tcp 54.146.60.183:443: i/o timeout"
Jul 05 06:16:47 ip-10-0-49-25 release-image-download.sh[2156]: Error: initializing source docker://registry.build05.ci.openshift.org/ci-op-vljni94d/release@sha256:250b9972d03463ba38a85b657066aa0706bf348608797aa50ddda769d4afdf27: pinging container registry registry.build05.ci.openshift.org: Get "https://registry.build05.ci.openshift.org/v2/": dial tcp 54.158.159.119:443: i/o timeout
Jul 05 06:16:47 ip-10-0-49-25 release-image-download.sh[2000]: Pull failed. Retrying registry.build05.ci.openshift.org/ci-op-vljni94d/release@sha256:250b9972d03463ba38a85b657066aa0706bf348608797aa50ddda769d4afdf27...
```
The installation is expecting to pull payload image from registry.build05.ci.openshift.org which is resolved from `release:latest` image stream in prow build farm, while oc-mirror only support mirrored payload images from quay.io.

Mirror [log](https://gcsweb-qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/qe-private-deck/pr-logs/pull/openshift_release/54021/rehearse-54021-periodic-ci-openshift-openshift-tests-private-release-4.16-arm64-stable-aws-ipi-disc-priv-sno-by-oc-mirror-f14/1809097345938231296/artifacts/aws-ipi-disc-priv-sno-by-oc-mirror-f14/mirror-images-by-oc-mirror/build-log.txt):
```
imageContentSources:
  - mirrors:
      - ec2-54-80-52-29.compute-1.amazonaws.com:5000/ci-op-vljni94d/release/openshift/release
    source: quay.io/openshift-release-dev/ocp-v4.0-art-dev
  - mirrors:
      - ec2-54-80-52-29.compute-1.amazonaws.com:5000/ci-op-vljni94d/release/openshift/release-images
    source: quay.io/openshift-release-dev/ocp-release       ---> Installer tried to pull image from registry.build05.ci.openshift.org/ci-op-vljni94d/release
```

This is a limit from oc-mirror, because it does not support to mirror a specific customer payload image, while `oc adm release mirror` does not have such problems.

**Comment by jianlinliu** [comment]
Though because of oc-mirror [limitation](https://github.com/openshift/release/pull/54021#issuecomment-2210448790), we can not run integration ci jobs for it, but the basic oc-mirror step works well. So I prefer to merge this PR, we can reference it in the future if needed.

cc @yunjiang29 and @gpei to review.

**Comment by gpei** [comment]
> The installation is expecting to pull payload image from registry.build05.ci.openshift.org which is resolved from release:latest image stream in prow build farm, while oc-mirror only support mirrored payload images from quay.io.

@jianlinliu could we manually replace the `quay` with `registry.build05.ci.openshift.org` in the imageContentSources to workaround this?

File: ci-operator/step-registry/mirror-images/by-oc-mirror/mirror-images-by-oc-mirror-commands.sh [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    fi
+    ls "${target_dir}/oc-mirror" || return 2
+    chmod +x "${target_dir}/oc-mirror" || return 2
+    return 0
+}
+
+# private mirror registry host
+# <public_dns>:<port>
+MIRROR_REGISTRY_HOST=$(head -n 1 "${SHARED_DIR}/mirror_registry_url")
+if [ ! -f "${SHARED_DIR}/mirror_registry_url" ]; then
```
[the remark this answers is not in the document]
Comment (jianlinliu) [reply ci-operator/step-registry/mirror-images/by-oc-mirror/mirror-images-by-oc-mirror-commands.sh:63]: Ah, good catch.

[the remark this answers is not in the document]
Comment (jianlinliu) [reply ci-operator/step-registry/mirror-images/by-oc-mirror/mirror-images-by-oc-mirror-commands.sh]: I delete the `if` block, `errexit` can catch it, if the file does not exist.

[the remark this answers is not in the document]
Comment (gpei) [reply ci-operator/step-registry/mirror-images/by-oc-mirror/mirror-images-by-oc-mirror-commands.sh]: yes, indeed

**Comment by yunjiang29** [comment]
lgtm

**Pull request merged**